core/src/main/scala/api/sfs.scala

```diff
@@ -25,15 +30,19 @@ trait Filesystem {
    *  things too eagerly, nor caching data for too long.
    */
   def resolve(path: Path): Key
-  def metadata(key: Key): Metadata
-  def lookup(key: Key): Data
+  def metadata(key: Key): M[Metadata]
+  def lookup(key: Key): M[Data]
```

Comment:
Oh, I just noticed that the first sentence of my reply might be wrong. I do like the single method approach, I don't think it should be effect free. I think it should be: `def resolve(path: Path): M[Metadata]`